### PR TITLE
Versioned acceleration structures

### PR DESCRIPTION
**Pull request by VZout**
Gives about a 6 fps increase on sponza.

### CODE REVIEW

File: src/d3d12/d3d12_acceleration_structure..cpp [hunk 1]
```diff
@@ -176,39 +291,31 @@ namespace wr::d3d12
...
-		{
-			BuildAccelerationStructure(cmd_list->m_native);
-		}
-		else if (GetRaytracingType(device) == RaytracingType::FALLBACK)
+		for (std::uint8_t i = 1; i < settings::num_back_buffers; i++)
 		{
-			// Set the descriptor heaps to be used during acceleration structure build for the Fallback Layer.
-			d3d12::BindDescriptorHeap(cmd_list, desc_heap, desc_heap->m_create_info.m_type, 0, true); //TODO: note this non frame idx
-			BuildAccelerationStructure(cmd_list->m_native_fallback);
+			//internal::CopyInstDescResource(cmd_list, blas, 0, i);
```
Comment (tntmeijs) [inline src/d3d12/d3d12_acceleration_structure..cpp]: Old code? If so, remove.

File: src/d3d12/d3d12_acceleration_structure..cpp [hunk 2]
```diff
@@ -343,16 +385,32 @@ namespace wr::d3d12
...
 
 	void UpdateTopLevelAccelerationStructure(AccelerationStructure& tlas, Device* device,
 		CommandList* cmd_list,
 		DescriptorHeap* desc_heap,
-		std::vector<desc::BlasDesc> blas_list)
+		std::vector<desc::BlasDesc> blas_list,
+		std::uint32_t frame_idx)
 	{
-		D3D12_RAYTRACING_ACCELERATION_STRUCTURE_BUILD_FLAGS build_flags = D3D12_RAYTRACING_ACCELERATION_STRUCTURE_BUILD_FLAG_PERFORM_UPDATE;
+		//D3D12_RAYTRACING_ACCELERATION_STRUCTURE_BUILD_FLAGS build_flags = D3D12_RAYTRACING_ACCELERATION_STRUCTURE_BUILD_FLAG_PERFORM_UPDATE;
```
Comment (tntmeijs) [inline src/d3d12/d3d12_acceleration_structure..cpp]: Does this still need to be here?

File: src/d3d12/d3d12_acceleration_structure..cpp [hunk 3]
```diff
@@ -362,129 +420,56 @@ namespace wr::d3d12
...
-			if (tlas.m_scratch)
-			{
-				//tlas.m_scratch->Release();
-			}
-
-			// Allocate scratch resource
-
-			tlas.m_rebuild_scratch = false;
-		}
+		//internal::UpdatePrebuildInfo(device, tlas, top_level_inputs);
```
Comment (tntmeijs) [inline src/d3d12/d3d12_acceleration_structure..cpp]: Are you sure this comment needs to be in the codebase?

File: src/d3d12/d3d12_acceleration_structure..cpp [hunk 4]
```diff
@@ -362,129 +420,56 @@ namespace wr::d3d12
...
 
 		bool rebuild_accel_structure = old_prebuild_info.ResultDataMaxSizeInBytes < tlas.m_prebuild_info.ResultDataMaxSizeInBytes ||
 			old_prebuild_info.ScratchDataSizeInBytes < tlas.m_prebuild_info.ScratchDataSizeInBytes ||
 			old_prebuild_info.UpdateScratchDataSizeInBytes < tlas.m_prebuild_info.UpdateScratchDataSizeInBytes;
 
-		if (rebuild_accel_structure)
+		if (false)
```
Comment (tntmeijs) [inline src/d3d12/d3d12_acceleration_structure..cpp]: I don't like this.

Comment (VZout) [reply src/d3d12/d3d12_acceleration_structure..cpp]: its debug code, good catch.

File: src/render_tasks/d3d12_path_tracer.hpp [hunk 5]
```diff
@@ -218,16 +217,10 @@ namespace wr
 			}
 
 			// Wait for AS to be built
-			{
-				auto barrier = CD3DX12_RESOURCE_BARRIER::UAV(as_build_data.out_tlas.m_native);
-				cmd_list->m_native->ResourceBarrier(1, &barrier);
-			}
+			d3d12::UAVBarrierAS(cmd_list, as_build_data.out_tlas, frame_idx);
```
Comment (LaisoEmilio) [inline src/render_tasks/d3d12_path_tracer.hpp]: double UAV Barrier? (line 203 already has a barrier, maybe my misunderstanding?)

File: src/d3d12/d3d12_acceleration_structure..cpp (hunk 4 above)
Comment (LaisoEmilio) [inline src/d3d12/d3d12_acceleration_structure..cpp]: meh

File: src/d3d12/d3d12_acceleration_structure..cpp [hunk 6]
```diff
@@ -343,16 +385,32 @@ namespace wr::d3d12
...
+		{
+			SAFE_RELEASE(as);
+		}
+		for (auto& inst_desc : structure.m_instance_descs)
+		{
+			SAFE_RELEASE(inst_desc);
+		}
+	}
+
+	void UAVBarrierAS(CommandList* cmd_list, AccelerationStructure structure, std::uint32_t frame_idx)
```
Comment (LaisoEmilio) [inline src/d3d12/d3d12_acceleration_structure..cpp]: can this be a reference to AccelerationStructure?

File: src/render_tasks/d3d12_path_tracer.hpp (hunk 5 above)
Comment (VZout) [reply src/render_tasks/d3d12_path_tracer.hpp]: your right, Just modified the existing code and didn't change anything. fixing now

File: src/d3d12/d3d12_acceleration_structure..cpp (hunk 4 above)
Comment (VZout) [reply src/d3d12/d3d12_acceleration_structure..cpp]: debug code, fixed

File: src/d3d12/d3d12_acceleration_structure..cpp (hunk 6 above)
Comment (VZout) [reply src/d3d12/d3d12_acceleration_structure..cpp]: No thanks to fallback, but can be const ref or I need to cast away const.

**Pull request merged**